service: Add optional Ports for CEC Services redirect

jrajahalme opened the pull request:
Add optional ports field into CEC/CCEC Services redirect, which, if not empty, will limit the redirected frontends to the ones whose port number is on the ports list.

Note that we also limit the frontend protocol to TCP, as we do not yet support other protocols in our Envoy listeners.

```release-note
CiliumEnvoyConfig CRDs now support an optional 'ports' field in services objects, limiting the redirected service frontends to the ones whose port is listed.
```

jrajahalme committed:
service: Add optional Ports for CEC Services redirect

Add optional ports field into CEC/CCEC Services redirect, which, if not
empty, will limit the redirected frontends to the ones whose port number
is on the ports list.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

mhofstetter commented on pkg/ciliumenvoyconfig/cec_manager.go:
```diff
@@ -127,7 +127,7 @@ func (r *cecManager) addK8sServiceRedirects(resourceName service.L7LBResourceNam
 
 		// Tell service manager to redirect the service to the port
```
nit: maybe enrich that comment that if configured, only a subset of frontend ports is forwarded to the L7 proxy.

jrajahalme replied:
leaving this for future enhancement

mhofstetter reviewed (approved):
LGTM. thanks!

learnitall reviewed (approved):
API changes look good, thanks!

julianwiedmann commented on pkg/service/service.go:
```diff
@@ -203,6 +203,31 @@ type L7LBInfo struct {
+// is empty.
+// 'ports' is typically short for no point optimizing the search.
+func (i *L7LBInfo) isPortMatch(fe *lb.L4Addr) bool {
+	// L7 LB redirect is only supported for TCP frontends
+	if fe.Protocol != lb.TCP {
+		return false
+	}
```
Can you please break this part out into a separate patch? Also, let's call this `isProtoAndPortMatch` or similar?

jrajahalme replied:
Done.

julianwiedmann reviewed (changes requested):
Just one small structural change request.

jrajahalme committed:
service: Only redirect TCP frontends to L7 LB

Limit the frontend protocol to TCP, as we do not yet support other
protocols in Envoy listeners.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

julianwiedmann reviewed (approved):
lgtm, thanks!
